### DISCUSSION TITLE
Making the exceptions pickleable

### DISCUSSION THREAD

**Discussion by randrej** [encode/httpx#1987] [Ideas]
I'm often using `httpx` with multiprocessing queues and such, and whenever `httpx` throws I get an exception similar to [this one](https://bugs.python.org/issue37208). Adding proper `__reduce__` methods to the exceptions, like explained [here](https://stackoverflow.com/a/36342588) would mitigate this issue completely.

Here's a concrete example of the error occurring:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/concurrent/futures/process.py", line 384, in wait_result_broken_or_wakeup
    result_item = result_reader.recv()
  File "/usr/local/lib/python3.9/multiprocessing/connection.py", line 256, in recv
    return _ForkingPickler.loads(buf.getbuffer())
TypeError: __init__() missing 1 required keyword-only argument: 'request'
```

I'm offering to contribute this functionality if the maintainers are willing to merge it.

**Comment by achille-roussel** [comment]
Hello!

I'm chiming in to support a resolution of this issue, httpx is a popular package and we often end up with pickle errors when exceptions are serialized.

I'm happy to send a pull request to address if there is consensus.

**Comment by achille-roussel** [comment]
A pull request was opened to address the issue https://github.com/encode/httpx/pull/3108

**Comment by camillol** [comment]
@tomchristie could you take another look at https://github.com/encode/httpx/pull/3108? Looks like it addressed all comments.
